openshift/odo
# url list and describe for docker

Signed-off-by: Stephanie <stephanie.cao@ibm.com>

**What type of PR is this?**
> Uncomment only one ` /kind` line, and delete the rest.
> For example, `> /kind bug` would simply become: `/kind bug`


/kind feature
/area url


**What does does this PR do / why we need it**:
This PR implement `url list` and `url describe` for local scenarios, supports output as json format as well, by specifying `-o json`

This PR also include integration tests for `url list` and `url describe` for docker. 

**Which issue(s) this PR fixes**:

Fixes #2934 

**How to test changes / Special notes to the reviewer**:
Automation test: 
simply run `make test-cmd-docker-devfile-url`

Manual test: 
```
odo preference set experimental true
odo preference set pushtarget docker
odo create java-spring-boot
odo url create testurl1 --now
```
Then run `url list`, should see testurl1's state is `Pushed`, output as json format also works. 
```
$ odo url list
Found the following URLs for component java-spring-boot
NAME         STATE          URL                 PORT
testurl1     Pushed         127.0.0.1:60574     8080

$ odo url list -o json
{
    "kind": "List",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {},
    "items": [
        {
            "kind": "url",
            "apiVersion": "odo.openshift.io/v1alpha1",
            "metadata": {
                "name": "testurl2",
                "creationTimestamp": null
            },
            "spec": {
                "host": "127.0.0.1",
                "port": 8080,
                "secure": false,
                "externalport": 60574
            },
            "status": {
                "state": "Pushed"
            }
        }
    ]
}
```

Then delete the testurl1 and run `url list`, should see the state becomes `Locally Deleted`
```
$ odo url delete testurl1
? Are you sure you want to delete the url testurl1 Yes
 ✓  URL testurl1 removed from the env file

To delete the URL on the cluster, please use `odo push`
$ odo url list
Found the following URLs for component java-spring-boot
NAME         STATE               URL                 PORT
testurl1     Locally Deleted     127.0.0.1:60574     8080

There are local changes. Please run 'odo push'.
``` 

Run `odo push` to unassign the exposed port from container. And run `url create` again, but do not push.  Then run `url list`, should see the state becomes `Not Pushed`
```
$ odo url list
NAME         STATE          URL                 PORT
testurl1     Not Pushed     ://     8080

```

Test will be similar for `url describe`
```
$ odo url create testurl2
 ✓  URL testurl2 created for component: java-spring-boot with exposed port: 49201

To apply the URL configuration changes, please use `odo push`
$ odo url describe testurl2
NAME         STATE          URL                 PORT
testurl2     Not Pushed     ://     8080

There are local changes. Please run 'odo push'.
$ odo url describe testurl2 -o json
{
    "kind": "url",
    "apiVersion": "odo.openshift.io/v1alpha1",
    "metadata": {
        "name": "testurl2",
        "creationTimestamp": null
    },
    "spec": {
        "host": "://",
        "port": 8080,
        "secure": false,
        "externalport": 49201
    },
    "status": {
        "state": "Not Pushed"
    }
}

$ odo push

Validation
 ✓  Validating the devfile [146390ns]

Creating Docker resources for component java-spring-boot
 •  Updating the component runtime  ...
 ✓  URL 127.0.0.1:49201 created
 ✓  Starting container for maysunfaisal/springbootruntime [401ms]
 ✓  Updating the component runtime [646ms]

Syncing to component java-spring-boot
 ✓  Checking files for pushing [3ms]
 ✓  Syncing files to the component [219ms]

Executing devfile commands for component java-spring-boot
 ✓  Executing devbuild command "/artifacts/bin/build-container-full.sh" [2s]
 ✓  Executing devrun command "/artifacts/bin/start-server.sh" [1s]

Pushing devfile component java-spring-boot
 ✓  Changes successfully pushed to component

$ odo url describe testurl2
NAME         STATE      URL                 PORT
testurl2     Pushed     127.0.0.1:49201     8080

$ odo url delete testurl2
? Are you sure you want to delete the url testurl2 Yes
 ✓  URL testurl2 removed from the env file

To delete the URL on the cluster, please use `odo push`

$ odo url describe testurl2
NAME         STATE               URL                 PORT
testurl2     Locally Deleted     127.0.0.1:49201     8080

There are local changes. Please run 'odo push'.
```
